src/core/internal/hash.d

```diff
@@ -173,21 +173,21 @@ nothrow pure @safe unittest // issue 19005
 
 //delegate hash. CTFE unsupported
 @trusted @nogc nothrow pure
-size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && is(T == delegate))
+size_t hashOf(T)(scope const T val, size_t seed = 0) if (!is(T == enum) && is(T == delegate))
 {
     assert(!__ctfe, "unable to compute hash of "~T.stringof);
     const(ubyte)[] bytes = (cast(const(ubyte)*)&val)[0 .. T.sizeof];
     return bytesHash(bytes, seed);
 }
 
 //class or interface hash. CTFE depends on toHash
-size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && is(T == interface) || is(T == class))
+size_t hashOf(T)(T val, size_t seed = 0) if (!is(T == enum) && is(T == interface) || is(T == class))
 {
     return hashOf(val ? (cast(Object)val).toHash() : 0, seed);
 }
 
 //associative array hash. CTFE depends on base types
-size_t hashOf(T)(auto ref T aa, size_t seed = 0) if (!is(T == enum) && __traits(isAssociativeArray, T))
+size_t hashOf(T)(T aa, size_t seed = 0) if (!is(T == enum) && __traits(isAssociativeArray, T))
```

Comment:
It's not as straightforward, but you can foreach on e.g. a `const int[int]`, is it worth adding const here if possible? You'd have to check if you can call `key.hashOf` and `val.hashOf`.

Reply:
IMO it's worth doing but since it involves more than just making trivial attribute changes I'd prefer to do it in a future PR.

>You'd have to check if you can call `key.hashOf` and `val.hashOf`.

It's a bit more involved than that. No matter what the type is I'll be able to call `hashOf(key)` and `hashOf(cast(const typeof(key)) key)` but they might not have the same result, for instance if the key has a non-const `toHash` function.

Reply:
I understand. It's concerning to have `hashOf` return different results on const and non-const structs regardless of whether it's in an AA.